### PR TITLE
Release to Main

### PR DESCRIPTION
**Pull request by jeremyfelder**

### CODE REVIEW

File: goicicle/curves/bls12377/ntt.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+	if ret != 0 {
+		fmt.Print("error interpolating")
+		return -1
+	}
+
+	return 0
+}
+
+func VecScalarAdd(in1_d, in2_d unsafe.Pointer, size int) int {
```
Comment (vhnatyk) [inline goicicle/curves/bls12377/ntt.go:156]: doesn't really belong here along with other APIs imho

File: goicicle/curves/bls12377/ntt.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	if dp == nil {
+		err = errors.New("nullptr returned from generating twiddles")
+		return unsafe.Pointer(nil), err
+	}
+
+	return unsafe.Pointer(dp), nil
+}
+
+// Reverses d_scalars in-place
+func ReverseScalars(d_scalars unsafe.Pointer, len int) (int, error) {
```
Comment (vhnatyk) [inline goicicle/curves/bls12377/ntt.go:97]: worth moving to utils

File: goicicle/curves/bls12377/ntt.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	"github.com/ingonyama-zk/icicle/goicicle"
+)
+
+// #cgo CFLAGS: -I./include/
+// #cgo CFLAGS: -I/usr/local/cuda/include
+// #cgo LDFLAGS: -L${SRCDIR}/../../ -lbls12_377
+// #include "ntt.h"
+import "C"
+
+const (
```
Comment (vhnatyk) [inline goicicle/curves/bls12377/ntt.go:33]: missed that on review :( - needs to be removed as well in all curves

File: goicicle/curves/bls12381/ntt.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	"github.com/ingonyama-zk/icicle/goicicle"
+)
+
+// #cgo CFLAGS: -I./include/
+// #cgo CFLAGS: -I/usr/local/cuda/include
+// #cgo LDFLAGS: -L${SRCDIR}/../../ -lbls12_381
+// #include "ntt.h"
+import "C"
+
+const (
```
Comment (vhnatyk) [inline goicicle/curves/bls12381/ntt.go:33]: missed that on review :( - needs to be removed as well in all curves

File: goicicle/curves/bls12381/ntt.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	if dp == nil {
+		err = errors.New("nullptr returned from generating twiddles")
+		return unsafe.Pointer(nil), err
+	}
+
+	return unsafe.Pointer(dp), nil
+}
+
+// Reverses d_scalars in-place
+func ReverseScalars(d_scalars unsafe.Pointer, len int) (int, error) {
```
Comment (vhnatyk) [inline goicicle/curves/bls12381/ntt.go:97]: to utils

File: goicicle/curves/bls12381/ntt.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+	if ret != 0 {
+		fmt.Print("error interpolating")
+		return -1
+	}
+
+	return 0
+}
+
+func VecScalarAdd(in1_d, in2_d unsafe.Pointer, size int) int {
```
Comment (vhnatyk) [inline goicicle/curves/bls12381/ntt.go:156]: move to vector API

File: goicicle/curves/bls12381/ntt.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+	if ret != 0 {
+		fmt.Print("error subtracting scalar vectors")
+		return -1
+	}
+
+	return 0
+}
+
+func ToMontgomery(d_scalars unsafe.Pointer, len int) (int, error) {
```
Comment (vhnatyk) [inline goicicle/curves/bls12381/ntt.go:186]: to utils? or dedicated convert.go API

File: goicicle/curves/bn254/ntt.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+	"github.com/ingonyama-zk/icicle/goicicle"
+)
+
+// #cgo CFLAGS: -I./include/
+// #cgo CFLAGS: -I/usr/local/cuda/include
+// #cgo LDFLAGS: -L${SRCDIR}/../../ -lbn254
+// #include "ntt.h"
+import "C"
+
+const (
```
Comment (vhnatyk) [inline goicicle/curves/bn254/ntt.go:33]: remove for now

File: goicicle/curves/bn254/ntt.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+	if dp == nil {
+		err = errors.New("nullptr returned from generating twiddles")
+		return unsafe.Pointer(nil), err
+	}
+
+	return unsafe.Pointer(dp), nil
+}
+
+// Reverses d_scalars in-place
+func ReverseScalars(d_scalars unsafe.Pointer, len int) (int, error) {
```
Comment (vhnatyk) [inline goicicle/curves/bn254/ntt.go:97]: to utils

File: goicicle/curves/bn254/ntt.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+	if ret != 0 {
+		fmt.Print("error interpolating")
+		return -1
+	}
+
+	return 0
+}
+
+func VecScalarAdd(in1_d, in2_d unsafe.Pointer, size int) int {
```
Comment (vhnatyk) [inline goicicle/curves/bn254/ntt.go:156]: to vector API file

File: goicicle/curves/bn254/ntt.go [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+	if ret != 0 {
+		fmt.Print("error subtracting scalar vectors")
+		return -1
+	}
+
+	return 0
+}
+
+func ToMontgomery(d_scalars unsafe.Pointer, len int) (int, error) {
```
Comment (vhnatyk) [inline goicicle/curves/bn254/ntt.go:186]: to utils

File: goicicle/templates/ntt/ntt.go.tmpl [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+	if ret != 0 {
+		fmt.Print("error adding scalar vectors")
+		return -1
+	}
+
+	return 0
+}
+
+func VecScalarSub(in1_d, in2_d unsafe.Pointer, size int) int {
```
Comment (vhnatyk) [inline goicicle/templates/ntt/ntt.go.tmpl:153]: conversions and utils better fit into corresponding API files

**Review by LeonHibnik — APPROVED** [review]

**Pull request merged**